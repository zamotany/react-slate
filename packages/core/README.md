<p align="center">
  <img alt="react-slate" src="https://raw.githubusercontent.com/zamotany/react-slate/master/docs/_assets/react_slate_logo.png" width="700">
</p>

<p align="center">
  Write interactive CLI apps with React
</p>

---

### Website & documentation

https://zamotany.github.io/react-slate/

### Features

* Render React apps to terminal (`render`), string (`renderToString`) or as a JSON tree (`renderToJson`).
* Build layouts with FlexBox (https://github.com/vislyhq/stretch).
* Render your app to alternative screen buffer and go back to the previous content upon exit.
* Log messages to main screen buffer with `console` when exiting.
* Disable colors with [`NO_COLOR`](https://no-color.org/) environment variable.
* Use standard ANSI colors, RBG, Hex and CSS keywords to style your app.
* Use built-in `<Progress />` and `<Spinner />` components.
* Get layout information using `onLayout` prop.
* Handle clicks using `onClick` prop.
* Detect mouse wheel events using `onWheel` prop.
* Use `for await` to continuously render your app using `renderToString` or `renderToJson`.
* Support for absolute positioning and depth (`zIndex`).
* Built with TypeScript.

Please check out [Roadmap](https://github.com/zamotany/react-slate/issues/99) for in-progress and planned features. 

### Limitations

* Web components are not compatible.
* Strings must be wrapped in a `<Text>` component.

## Installation

```bash
yarn add react @react-slate/core
```
