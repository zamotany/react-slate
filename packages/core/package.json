{
  "name": "@react-slate/core",
  "version": "1.0.0-alpha.2",
  "description": "Write interactive CLI apps with React",
  "license": "MIT",
  "repository": "https://github.com/zamotany/react-slate",
  "private": false,
  "author": {
    "name": "Paweł Trysła",
    "email": "zamotany.oss@gmail.com"
  },
  "main": "build/index.js",
  "files": [
    "/build",
    "/vendor",
    "index.js",
    "README.md"
  ],
  "engines": {
    "node": ">=10"
  },
  "publishConfig": {
    "access": "public"
  },
  "peerDependencies": {
    "react": ">=16.9.0"
  },
  "dependencies": {
    "ansi-escapes": "^4.2.1",
    "ansi-styles": "^4.1.0",
    "cli-spinners": "^2.2.0",
    "color-convert": "^2.0.1",
    "colorette": "^1.1.0",
    "fast-memoize": "^2.5.1",
    "react-reconciler": "0.22.2",
    "shallowequal": "^1.1.0",
    "supports-color": "^7.1.0",
    "strip-ansi": "^5.2.0",
    "terminal-kit": "^1.31.1"
  },
  "devDependencies": {
    "@types/ansi-styles": "^3.2.1",
    "@types/cli-spinners": "^1.3.0",
    "@types/color-convert": "^1.9.0",
    "@types/strip-ansi": "^5.2.1",
    "@types/supports-color": "^5.3.0",
    "@types/terminal-kit": "^1.28.0",
    "react": "^16.9.0"
  }
}
