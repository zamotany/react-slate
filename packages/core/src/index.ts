export { terminal } from 'terminal-kit';

export { default as renderToJson } from './renderToJson';
export { default as renderToString } from './renderToString';
export { default as renderFullscreen, exit } from './renderFullscreen';
export {
  default as render,
  withRenderStatic,
  useRenderStatic,
  RenderStatic,
} from './render';

export { default as View, Props as ViewProps } from './components/View';
export { default as Text, Props as TextProps } from './components/Text';
export { default as Progress } from './components/Progress';
export { default as Spinner } from './components/Spinner';

export { TextTransform, JsonText, JsonView, JsonParagraph } from './types';
