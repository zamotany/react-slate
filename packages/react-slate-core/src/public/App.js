/* @flow */

const onExitListeners = [[], [], [], []];
const onErrorListeners = [[], [], [], []];

const Priority = {
  Critical: 3,
  High: 2,
  Normal: 1,
  Low: 0,
};

function execListeners(listeners, ...args) {
  for (let i = Priority.Critical; i >= Priority.Low; i--) {
    // $FlowFixMe
    listeners[i].forEach(listener => listener(...args));
    listeners[i] = []; // eslint-disable-line no-param-reassign
  }
}

process.on('exit', code => {
  if (code === 0) {
    execListeners(onExitListeners);
  } else {
    execListeners(onErrorListeners);
  }
});

process.on('SIGINT', () => {
  execListeners(onExitListeners);
  process.exit(0);
});

process.on('uncaughtException', error => {
  execListeners(onErrorListeners, error);
  process.exitCode = 1;
});

function validatePriority(priority?: number) {
  if (
    typeof priority === 'number' &&
    (priority < Priority.Low || priority > Priority.Critical)
  ) {
    throw new Error();
  }
}

function removeListener(listeners, listener, priority) {
  if (typeof priority === 'number') {
    const index = listeners.indexOf(listener);
    if (index >= 0) {
      listeners.splice(index, 1);
    }
  } else {
    onErrorListeners.forEach(_listeners => {
      const index = _listeners.indexOf(listener);
      if (index >= 0) {
        _listeners.splice(index, 1);
      }
    });
  }
}

export default {
  Priority,
  get exitCode() {
    return process.exitCode;
  },
  set exitCode(code: number) {
    process.exitCode = code;
  },
  exit(code: number = 0) {
    process.exit(code);
  },
  onExit(listener: () => void, priority: number = Priority.Normal) {
    validatePriority(priority);
    if (onExitListeners[priority].indexOf(listener) === -1) {
      onExitListeners[priority].push(listener);
    }
  },
  onError(
    listener: (error?: Error) => void,
    priority: number = Priority.Normal
  ) {
    validatePriority(priority);
    if (onErrorListeners[priority].indexOf(listener) === -1) {
      onErrorListeners[priority].push(listener);
    }
  },
  removeOnExit(listener: () => void, priority?: number) {
    validatePriority(priority);
    removeListener(onExitListeners, listener, priority);
  },
  removeOnError(listener: (error?: Error) => void, priority?: number) {
    validatePriority(priority);
    removeListener(onErrorListeners, listener, priority);
  },
};
