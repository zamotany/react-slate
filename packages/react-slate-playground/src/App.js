import React from 'react';
import { View, Spinner, ProgressBar, KeyPress } from 'react-slate';

const components = ['Spinner', 'ProgressBar'];

export default class App extends React.Component {
  state = {
    componentPreview: 0,
    progress: 0,
    isMoving: false,
    previewLeftOffset: 0,
    previewTopOffset: 0,
  };

  componentDidMount() {
    setInterval(() => {
      this.setState(state => ({
        progress:
          state.progress >= 1 ||
          state.componentPreview !== components.indexOf('ProgressBar')
            ? 0
            : state.progress + 0.1,
      }));
    }, 100);
  }

  onPress = char => {
    if (!this.state.isMoving && char === 'w') {
      this.setState(state => ({
        componentPreview: Math.max(0, state.componentPreview - 1),
      }));
    } else if (!this.state.isMoving && char === 's') {
      this.setState(state => ({
        componentPreview: Math.min(
          components.length - 1,
          state.componentPreview + 1
        ),
      }));
    } else if (char === 'm') {
      this.setState(state => ({ isMoving: !state.isMoving }));
    } else if (this.state.isMoving) {
      let dx = 0;
      let dy = 0;
      switch (char) {
        case 'w':
          dy -= 1;
          break;
        case 'a':
          dx -= 1;
          break;
        case 's':
          dy += 1;
          break;
        case 'd':
          dx += 1;
          break;
        default:
      }
      this.setState(state => ({
        previewLeftOffset: state.previewLeftOffset + dx,
        previewTopOffset: state.previewTopOffset + dy,
      }));
    }
  };

  renderPreview(componentIndex) {
    switch (components[componentIndex]) {
      case 'Spinner':
        return (
          <View>
            <Spinner style={styles.spinner} />
            <Spinner style={styles.spinner} type="line" />
            <Spinner type="dots11" />
          </View>
        );
      case 'ProgressBar':
        return <ProgressBar barWidth={16} value={this.state.progress} />;
      default:
        return null;
    }
  }

  render() {
    return (
      <View style={styles.appContainer}>
        <View style={styles.intro}>react-slate playground</View>
        <View style={styles.menu}>
          <View style={{ ...styles.menuInfo, marginBottom: 0 }}>
            {'(use "w" and "s" to navigate)'}
          </View>
          <View style={styles.menuInfo}>
            {'(press "m" to move preview using "w", "a", "s", "d")'}
          </View>
          {components.map((name, index) => (
            <View
              key={name}
              style={{
                ...styles.menuItem,
                ...(index === this.state.componentPreview &&
                  styles.menuItemActive),
              }}
            >
              {name}
            </View>
          ))}
        </View>
        <View
          style={{
            ...styles.componentPreview,
            left: 22 + this.state.previewLeftOffset,
            top: 5 + this.state.previewTopOffset,
            borderColor: this.state.isMoving ? 'ansi-magenta' : 'ansi-white',
          }}
        >
          <View style={styles.componentPreviewLabel}>Preview:</View>
          {this.renderPreview(this.state.componentPreview)}
        </View>
        <KeyPress stream={process.stdin} onPress={this.onPress} />
      </View>
    );
  }
}

const styles = {
  appContainer: {
    marginLeft: 2,
  },
  intro: {
    color: 'ansi-green',
    margin: '0 0 1 0',
  },
  menu: {
    marginBottom: 1,
  },
  menuInfo: {
    marginBottom: 1,
    color: 'ansi-gray',
  },
  menuItem: {
    width: 13,
    paddingLeft: 1,
  },
  menuItemActive: {
    backgroundColor: 'ansi-blue',
  },
  componentPreview: {
    borderStyle: 'solid',
    width: 20,
    height: 6,
    position: 'fixed',
    textAlign: 'center',
  },
  componentPreviewLabel: {
    marginBottom: 1,
  },
  spinner: {
    marginRight: 1,
  },
};
