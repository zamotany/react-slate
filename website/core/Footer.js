/**
 * Copyright (c) 2017-present, Facebook, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

const React = require('react');

class Footer extends React.Component {
  docUrl(doc, language) {
    const baseUrl = this.props.config.baseUrl;
    return baseUrl + 'docs/' + (language ? language + '/' : '') + doc;
  }

  pageUrl(doc, language) {
    const baseUrl = this.props.config.baseUrl;
    return baseUrl + (language ? language + '/' : '') + doc;
  }

  render() {
    const currentYear = new Date().getFullYear();
    return (
      <footer className="nav-footer" id="footer">
        <section className="sitemap">
          <div>
            <a href={this.props.config.repoUrl}>GitHub</a>
            <a
              className="github-button"
              href={this.props.config.repoUrl}
              data-icon="octicon-star"
              data-count-href="/zamotany/react-slate/stargazers"
              data-show-count={true}
              data-count-aria-label="# stargazers on GitHub"
              aria-label="Star this project on GitHub">
              Star
            </a>
          </div>
        </section>
      </footer>
    );
  }
}

module.exports = Footer;
