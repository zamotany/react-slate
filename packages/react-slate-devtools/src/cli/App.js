/* @flow */

import React from 'react';
import { View } from '@react-slate/core';
import Profiler from './Profiler';

type State = {
  status: 'connected' | 'disconnected',
  socket: ?WebSocket,
};

type Props = {
  server: *,
};

export default class App extends React.Component<Props, State> {
  state = {
    status: 'disconnected',
    socket: null,
  };

  constructor(props: Props) {
    super(props);

    this.props.server.on('connection', socket => {
      this.setState({ status: 'connected', socket });
      socket.on('close', () => {
        this.setState({ status: 'disconnected', socket: null });
      });
      const intervalId = setInterval(() => {
        try {
          socket.send('ping');
        } catch (error) {
          socket.close();
          this.setState({ status: 'disconnected', socket: null });
          clearInterval(intervalId);
        }
      }, 100);
    });
  }

  render() {
    return (
      <View>
        <View style={[styles.title, styles.divider]}>devtools</View>
        <View style={styles.divider}>Status: {this.state.status}</View>
        <Profiler socket={this.state.socket} />
      </View>
    );
  }
}

const styles = {
  title: {
    color: 'black',
    backgroundColor: 'green',
    textTransform: 'uppercase',
    padding: '0 2',
  },
  divider: {
    marginBottom: 1,
  },
};
