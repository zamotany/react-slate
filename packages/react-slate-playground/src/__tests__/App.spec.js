import React from 'react';
import { renderToString } from '@react-slate/core';
import stripAnsi from 'strip-ansi';
import App from '../App';

test('App should render', () => {
  expect(stripAnsi(renderToString(<App />, { width: 60 }))).toMatchSnapshot();
});
