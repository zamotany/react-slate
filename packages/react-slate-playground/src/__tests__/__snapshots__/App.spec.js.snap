// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should render 1`] = `
"  react-slate playground                                    
                                                            
  (use \\"w\\" and \\"s\\" to navigate)                             
  (press \\"m\\" to move preview using \\"w\\", \\"a\\", \\"s\\", \\"d\\")      
                                                            
   Spinner                    ┌──────────────────┐          
   ProgressBar                │     Preview:     │          
                              │                  │          
  ┌───────┐                   │                  │          
  │Scroll!│                   │                  │          
  └───────┘                   └──────────────────┘          
  Press \\"x\\" to disable scroll                               
                                                            
                                                            
                                                            
                                                            
                                                            
                                                            
                                                            
                                                            "
`;
